File: SRAudioKit/SRAudioDeviceManager.m
//
//  SRAudioDeviceManager.m
//  SRAudioKitDemoForOSX
//
//  Created by Heeseung Seo on 2015. 2. 9..
//  Copyright (c) 2015년 Seorenn. All rights reserved.
//

#import "SRAudioDeviceManager.h"
#import "SRAudioDevice.h"

#pragma mark - C Implementations

static OSStatus getAudioDevices(Ptr *devices, UInt16 *devicesAvailable) {
    OSStatus err = noErr;
    UInt32 dataSize = 0;
    
    AudioObjectPropertyAddress address = {
        kAudioHardwarePropertyDevices,
        kAudioObjectPropertyScopeGlobal,
        kAudioObjectPropertyElementMaster
    };
    
    err = AudioObjectGetPropertyDataSize(kAudioObjectSystemObject, &address, 0, NULL, &dataSize);
    if (err != noErr) return err;
    
    *devicesAvailable = dataSize / (UInt32)sizeof(AudioObjectID);
    if (*devicesAvailable < 1) {
        // No Devices
        return err;
    }
    
    if (*devices != NULL) free(*devices);
    *devices = (Ptr)malloc(dataSize);
    memset(*devices, 0, dataSize);
    
    err = AudioObjectGetPropertyData(kAudioObjectSystemObject, &address, 0, NULL, &dataSize, (void *)*devices);
    if (err != noErr) {
        free(*devices);
    }
    
    return err;
}

#pragma mark - Objective-C Class Implementations

@implementation SRAudioDeviceManager

+ (NSArray *)devices {
    AudioObjectID *devices = NULL;
    UInt16 devicesAvailable = 0;
    
    NSMutableArray *deviceArray = [[NSMutableArray alloc] init];
    
    if (getAudioDevices((Ptr *)&devices, &devicesAvailable) != noErr) {
        return nil;
    }
    
    for (int i=0; i < devicesAvailable; i++) {
        UInt16 deviceID = devices[i];
        
        SRAudioDevice *device = [[SRAudioDevice alloc] initWithDeviceID:deviceID];
        [deviceArray addObject:device];
    }
    
    free(devices);
    
    return [deviceArray copy];
}

+ (SRAudioDeviceManager *)sharedManager {
    static dispatch_once_t onceToken;
    static SRAudioDeviceManager *instance = nil;
    dispatch_once(&onceToken, ^{
        instance = [[SRAudioDeviceManager alloc] init];
    });
    return instance;
}

@end

File: SRAudioKit/Sources/SRAudioDeviceManager.swift
//
//  SRAudioDeviceManager.swift
//  SRAudioKit
//
//  Created by Heeseung Seo on 2015. 9. 23..
//  Copyright © 2015년 Seorenn. All rights reserved.
//

import Foundation
import SRAudioKitPrivates
import AudioToolbox
import CoreAudio

open class SRAudioDeviceManager {
    open static let sharedManager: SRAudioDeviceManager = SRAudioDeviceManager()

    open var devices: [SRAudioDevice] {
        #if os(OSX)
            var dataSize: UInt32 = 0;
            var results = [SRAudioDevice]()
            
            var address = AudioObjectPropertyAddress(
                mSelector: kAudioHardwarePropertyDevices,
                mScope: kAudioObjectPropertyScopeGlobal,
                mElement: kAudioObjectPropertyElementMaster)
            
            var err = AudioObjectGetPropertyDataSize(AudioObjectID(kAudioObjectSystemObject), &address, 0, nil, &dataSize);
            
            guard err == noErr else { return results }
            
            let count = Int(dataSize / UInt32(MemoryLayout<AudioObjectID>.size))
            guard count > 0 else { return results }

            let devicesPtr = UnsafeMutablePointer<AudioObjectID>.allocate(capacity: Int(dataSize))
            defer { devicesPtr.deallocate(capacity: Int(dataSize)) }
            
            err = AudioObjectGetPropertyData(AudioObjectID(kAudioObjectSystemObject), &address, 0, nil, &dataSize, devicesPtr);

            guard err == noErr else { return results }
            
            var curPtr = devicesPtr
            for _ in 0..<count {
                let deviceID = curPtr.pointee
                let d = SRAudioDevice(deviceID: deviceID)
                results.append(d)
                
                curPtr = curPtr.successor()
            }
            
            return results
        #else
            return [SRAudioDevice]()
        #endif
    }
    
    open var defaultInputDevice: SRAudioDevice? {
        #if os(OSX)
            var size = UInt32(MemoryLayout<AudioDeviceID>.size)
            var deviceID = AudioDeviceID()
            var address = AudioObjectPropertyAddress(
                mSelector: kAudioHardwarePropertyDefaultInputDevice,
                mScope: kAudioObjectPropertyScopeGlobal,
                mElement: kAudioObjectPropertyElementMaster)
            
            let error = AudioObjectGetPropertyData(AudioObjectID(kAudioObjectSystemObject), &address, 0, nil, &size, &deviceID)
            guard error == noErr else { return nil }
            
            return SRAudioDevice(deviceID: deviceID)
        #else
            return nil
        #endif
    }
    
    public init() {
        
    }

    
}
